pagefaultgames/pokerogue
# [Bug][Move] Refactor moves that call a random move

Rebranch of #3075 

This includes the moves: Metronome, Assist, Sleep Talk, Copycat, Mirror Move

## TODO
- [x] Prevent allies being targets of Metronome/Copycat/other random move targetting
  - [x] Verify that self-status/Aromatic Mist still can target allies
- [x] Fix issue of two-turn moves with Metronome/multi-hit moves only hitting once with Sleep Talk 
  - [x] Test refactor for removing QueuedMove in favor of TurnMove (see https://github.com/pagefaultgames/pokerogue/pull/3380/commits/714211d3817b04cb3d4e94d07e7996b15e22b741)
- [x] Refactor Copycat/Mirror Move. (potentially make a CallMoveAttr, have all other attributes extend off of this one)
- [x] Refactor Promises into async/await pattern
- [x] Write unit tests for 
  - [x] Metronome
  - [x] Sleep Talk
  - [x] Copycat
  - [x] Mirror Move

## What are the changes?
Properly implements Assist. 
Refactors RandomMovesetMoveAttr to extend RandomMoveAttr. 
Creates arrays of invalid moves for each move: Assist, Sleep Talk, Metronome, since they each have different sets of uncallable moves.

## Why am I doing these changes?
Assist did not call ally moves, had useless code, and was given no distinction in logic from Sleep Talk. 
Metronome calls random moves in a similar manner to the two other moves, so it makes sense to have Assist/Sleep Talk reuse the code for calling moves.
None of Assist/Sleep Talk/Metronome had properly implemented lists of uncallable moves.

## What did change?
- Assist now uses the entire team's party
- Assist/Sleep Talk's attribute now extends `RandomMoveAttr` and uses its callMoves function (which has been modified slightly to accommodate this) to call the move
- Sleep Talk targeting is changed to `NEAR_ENEMY`
- Copycat's last move used is now unaffected by moves that fail, see Copycat's [bulbapedia](https://bulbapedia.bulbagarden.net/wiki/Copycat_(move)#:~:text=The%20last%20move%20used%20is%20now%20unaffected%20if%20a%20Pok%C3%A9mon%20tries%20to%20use%20a%20move%20but%20is%20prevented%20by%20a%20status%20condition%20or%20effect.%20In%20other%20words%2C%20the%20last%2Dused%20move%20remains%20what%20it%20was%20before%20the%20Pok%C3%A9mon%27s%20turn.)
- Mirror Move can now properly copy status moves
- Mirror Move now targets a pokemon and copies that pokemon's last used move
- Assist/Sleep Talk/Metronome/Copycat now have individual arrays listing uncallable moves. These arrays were created referencing Bulbapedia's list of uncallable moves.
- Uncallable moves for [Assist](https://bulbapedia.bulbagarden.net/wiki/Assist_(move)#:~:text=in%20a%20battle.-,Uncallable%20moves,-This%20move%20effect)
- Uncallable moves for [Sleep Talk](https://bulbapedia.bulbagarden.net/wiki/Sleep_Talk_(move)#:~:text=Z%2DPower%20effect.-,Unselectable%20moves,-This%20section%20is)
- Uncallable moves for [Metronome](https://bulbapedia.bulbagarden.net/wiki/Metronome_(move)#:~:text=Z%2DPower%20effect.-,Unselectable%20moves,-This%20move%20effect)
- Uncallable moves for [Copycat](https://bulbapedia.bulbagarden.net/wiki/Copycat_(move)#:~:text=or%20failing%20entirely.-,Uncallable%20moves,-This%20move%20effect)

Note: Sleep Talk specifies that it targets the [opposite opponent](https://bulbapedia.bulbagarden.net/wiki/Sleep_Talk_(move)#:~:text=In%20Double%20and%20Triple%20Battles%2C%20because%20of%20the%20inability%20to%20select%20a%20target%2C%20the%20chosen%20move%20will%20always%20hit%20the%20opponent%20directly%20across%20from%20the%20user%2C%20if%20the%20move%20called%20normally%20hits%20only%20one%20Pok%C3%A9mon.). Since there is no `MoveTarget` specifically for opposite opponents, I chose to make it `NEAR_ENEMY` for now.

### Screenshots/Videos
<details>
  <summary>Metronome</summary>

https://github.com/user-attachments/assets/c8861801-acd0-4ae0-b817-181cffa549d7

</details>
<details>
  <summary>Assist</summary>

https://github.com/user-attachments/assets/9146fc45-9294-4d59-a726-ffcdf31b5777

</details>
<details>
  <summary>Sleep Talk</summary>

https://github.com/user-attachments/assets/afd243b0-b2e9-4d84-9ba9-13a924c4b439

</details>

## How to test the changes?
Use src/overrides.ts to give yourself one of the three moves. If using Metronome, observe it will never call certain moves. If using Assist, observe it calling your party pokemon's moves. If using Sleep Talk, observe it only calling its own moveset.
Note that each move will no longer call certain moves: attempt Assist/Sleep Talk with 2 turn moves to see them fail. 
List of moves that won't be callable are given above by links to bulbapedia.

## Checklist
